docs/quick-start/vpn-config.md
# Configuring VPN

## SSH tunnel with sshuttle

1. Get IP address of ingress service. On your AIO instance run following command:
```bash
kubectl -n openstack get svc ingress -o jsonpath='{.status.loadBalancer.ingress[].ip}'
```
2. Update local `/etc/hosts` file to point public domain to ingress external IP
```bash
10.172.1.100 aodh.it.just.works barbican.it.just.works cinder.it.just.works cloudformation.it.just.works designate.it.just.works glance.it.just.works gnocchi.it.just.works heat.it.just.works horizon.it.just.works keystone.it.just.works metadata.it.just.works neutron.it.just.works nova.it.just.works novncproxy.it.just.works octavia.it.just.works placement.it.just.works spiceproxy.it.just.works
```
where `10.172.1.100` is the result of previous command execution.
4. Setup `sshuttle` to services external IPs
```bash
sshuttle -r ubuntu@172.16.250.153 10.172.1.0/24
```
where `172.16.250.153` is the public IP of your AIO instance

## OpenVPN (for TryMOSK installation)

By default, the TryMOSK installation process sets up an OpenVPN server on the
instance and creates a client configuration file.

1. Copy OpenVPN client configuration file */src/vpn/client.ovpn* to your
local computer. Use `scp` or another secure file transfer method to download
the file from the instance to your local machine.

2. Update the server IP address in the configuration. In the `client.ovpn` file,
replace the placeholder `<Put your server public IP here>` (found on **line 4**)
with the actual **public IPv4 address** of your AWS instance.
**Example – before:**
```
...
proto udp
remote <Put your server public IP here> 1194
resolv-retry infinite
...
```
**Example – after:**
```
...
proto udp
remote 18.218.29.107 1194
resolv-retry infinite
...
```
Where `18.218.29.107` is the public IPv4 address of your AWS EC2 instance.

3. Use the configuration with an OpenVPN client such as:
    * The official OpenVPN client: [https://openvpn.net/client/](https://openvpn.net/client/)
    * Any compatible VPN client that supports OpenVPN connections

    Refer to your chosen VPN client’s documentation for import and connection instructions.

4. Configur DNS resolver. The OpenVPN server is configured to update the client’s
DNS resolver so that TryMOSK service URLs can be resolved correctly.
    * **macOS and Windows** – OpenVPN clients apply this configuration automatically.
    * **Linux** – Additional DNS configuration may be required, depending on your distribution,
to ensure proper domain resolution for TryMOSK services. Consult your distribution’s
networking documentation for details. You also can update `/etc/hosts` file on your
local computer as is described in step 2 for `sshuttle` setup.


docs/architecture/rockoon/configuration.md
# Configuration

The OpenStack Controller enables you to modify its configuration at runtime without restarting. MOSK stores the controller configuration
in the `openstack-controller-config` `ConfigMap` in the osh-system namespace of your cluster.

To retrieve the OpenStack Controller configuration `ConfigMap`, run:

```bash
kubectl get configmaps openstack-controller-config -o yaml
```

Example of OpenStackController configuration

```yaml
apiVersion: v1
data:
  extra_conf.ini: |
    [maintenance]
    respect_nova_az = false
kind: ConfigMap
metadata:
  annotations:
    openstackdeployments.lcm.mirantis.com/skip_update: "true"
  name: openstack-controller-config
  namespace: osh-system
```

```ini
[osctl]
# The number of seconds to wait for all component from application becomes ready
wait_application_ready_timeout = 1200

# The number of seconds to sleep between checking application ready attempts
wait_application_ready_delay = 10

# The amount of time to wit for flapping node
node_not_ready_flapping_timeout = 120

[helmbundle]
# The number of seconds to wait for values set in manifest are propagated to child objects.
manifest_enable_timeout = 600

# The number of seconds between attempts to check that values were applied.
manifest_enable_delay = 10

# The number of seconds to wait for values are removed from manifest and propagated to child objects.
manifest_disable_timeout = 600

# The number of seconds between attempts to check that values were removed from release.
manifest_disable_delay = 10

# The number of seconds to wait for kubernetes object removal
manifest_purge_timeout = 600

# The number of seconds between attempts to check that kubernetes object is removed
manifest_purge_delay = 10

# The number of seconds to pause for helmbundle changes
manifest_apply_delay = 10

# The number of seconds to run for helm command
helm_cmd_timeout = 120

[maintenance]
# number of instances to migrate concurrently
instance_migrate_concurrency = 1

# max number of compute nodes we allow to update in parallel
nwl_parallel_max_compute = 30

# max number of gateway nodes we allow to update in parallel
nwl_parallel_max_gateway = 1

# respect nova AZs, when set to true parallel update is allowed only for computes in same AZ
respect_nova_az = True

# flag to skip instance check on host before proceeding with node removal. By default is False
# which means that node removal will be blocked unless at least 1 instance exists on host.
ndr_skip_instance_check = False

# flag to skip volume check on host before proceeding with node removal. By default is False
# which means that node removal will be blocked unless at least 1 volume exists on host.
# Volume is tied to specific host only for LVM backend.
ndr_skip_volume_check = False
```


rockoon/settings.py
#    Copyright 2020 Mirantis, Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.
import configparser
import os
import glob
import random
import sys
import signal
import time
import faulthandler

import json
import kopf
from kopf._core.engines.posting import event_queue_var
from pathlib import Path

from rockoon import constants as const
from rockoon.utils import get_logger


faulthandler.register(signal.SIGHUP, all_threads=True)
faulthandler.enable(file=sys.stderr, all_threads=True)

HEARTBEAT = time.time()
CURRENT_NUMBER_OF_TASKS = -1
LOG = get_logger(__name__)


def bool_from_env(env_name, default):
    """Convert env variable into boolean
    :param env_name: the name of environment variable
    :param default: the default value to return
    :returns True: when value of env is in TRUE_STRINGS
    :returns False: when value of env is in FALSE_STRINGS
    :raise kopf.PermanentError: when value not in TRUE_STRINGS or FALSE_STRINGS
    """

    data = os.environ.get(env_name)

    if data is None:
        return default

    lowered = data.strip().lower()

    if lowered in const.TRUE_STRINGS:
        return True
    elif lowered in const.FALSE_STRINGS:
        return False

    raise kopf.PermanentError(f"Failed to convert {data} into boolean.")


def json_from_env(env_name, default):
    """Load a json string from an env variable

    :param env_name: the name of environment variable
    :param default: the default value to return
    :returns True: when value of env is in TRUE_STRINGS
    :returns False: when value of env is in FALSE_STRINGS
    :raise kopf.PermanentError: when value not in TRUE_STRINGS or FALSE_STRINGS

    """
    data = os.environ.get(env_name)

    if data is None:
        return default

    return json.loads(data)


class Config(configparser.ConfigParser):
    def __init__(self, *args, **kwargs):
        self.file_cache = {}
        self.conf_dir = "/etc/rockoon"
        self.filenames = self.get_config_files()
        super().__init__(*args, **kwargs)

    def __getitem__(self, item):
        self.read_config()
        return super().__getitem__(item)

    def _get(self, section, conv, option, **kwargs):
        self.read_config()
        return super()._get(section, conv, option, **kwargs)

    def get_config_files(self):
        res = []
        for path in [
            os.path.join(
                sys.prefix,
                "etc/rockoon",
            ),
            self.conf_dir,
            f"{self.conf_dir}/conf.d",
        ]:
            if os.path.isdir(path):
                for cfg in glob.glob(f"{path}/*.ini"):
                    res.append(cfg)
        return res

    def read_config(self):
        reloaded = True
        for conf_file in self.filenames:
            mtime = os.path.getmtime(conf_file)
            if mtime > self.file_cache.get("mtime", 0):
                reloaded = False
                self.file_cache["mtime"] = mtime
                break
        if not reloaded:
            LOG.info("Reloading configuration.")
            self.read(self.filenames)

    def getString(self, section, option):
        return self.get(section, option).strip('"')


# The name of openstack deployment namespace
OSCTL_OS_DEPLOYMENT_NAMESPACE = os.environ.get(
    "OSCTL_OS_DEPLOYMENT_NAMESPACE", "openstack"
)

# The name of os controller deployment namespace
OSCTL_CONTROLLER_NAMESPACE = os.environ.get(
    "OSCTL_CONTROLLER_NAMESPACE", "osh-system"
)

# The name of openstack deployment namespace
OSCTL_CEPH_SHARED_NAMESPACE = os.environ.get(
    "OSCTL_CEPH_SHARED_NAMESPACE", "openstack-ceph-shared"
)

# The location of clouds.yaml with admin credentials
OS_CLIENT_CONFIG_FILE = os.environ.get(
    "OS_CLIENT_CONFIG_FILE", "/etc/openstack/clouds.yaml"
)

# The name of cloud in clouds.yaml
OS_CLOUD = os.environ.get("OS_CLOUD", "osctl")
OS_CLOUD_SYSTEM = os.environ.get("OS_CLOUD_SYSTEM", f"{OS_CLOUD}-system")

# TODO(mkarpin): move openstack related settings to separate file
# as settings.py is imported inside kube.py
# Url for openstack binaries/helm charts
OSCTL_BINARY_BASE_URL = os.environ.get("OSCTL_BINARY_BASE_URL", "")
# Url for openstack docker images
OSCTL_IMAGES_BASE_URL = os.environ.get("OSCTL_IMAGES_BASE_URL", "")

# The name of secret with BGP information
OSCTL_BGPVPN_NEIGHBOR_INFO_SECRET_NAME = os.environ.get(
    "OSCTL_BGPVPN_BGP_NEIGHBOR_INFO_SECRET_NAME", "frr-bgp-neighbors"
)

# The number of retries while waiting a resouce deleted
OSCTL_RESOURCE_DELETED_WAIT_RETRIES = int(
    os.environ.get("OSCTL_RESOURCE_DELETED_WAIT_RETRIES", 120)
)

# The number of seconds to sleep while waiting a resouce deleted
OSCTL_RESOURCE_DELETED_WAIT_TIMEOUT = int(
    os.environ.get("OSCTL_RESOURCE_DELETED_WAIT_TIMEOUT", 1)
)

OSCTL_REDIS_NAMESPACE = os.environ.get(
    "OSCTL_REDIS_NAMESPACE", "openstack-redis"
)

OSCTL_HEARTBEAT_INTERVAL = int(os.environ.get("OSCTL_HEARTBEAT_INTERVAL", 300))

OSCTL_HEARTBEAT_MAX_DELAY = int(
    os.environ.get("OSCTL_HEARTBEAT_MAX_DELAY", OSCTL_HEARTBEAT_INTERVAL * 3)
)

# If we did not start applying change in OSCTL_APPLYING_MAX_DELAY seconds
# controller will be restarted
OSCTL_APPLYING_MAX_DELAY = int(
    os.environ.get("OSCTL_APPLYING_MAX_DELAY", "60")
)

OSCTL_BATCH_HEATH_UPDATER_PERIOD = int(
    os.environ.get("OSCTL_BATCH_HEATH_UPDATER_PERIOD", 60)
)

OSCTL_PYKUBE_HTTP_REQUEST_TIMEOUT = float(
    os.environ.get("OSCTL_PYKUBE_HTTP_REQUEST_TIMEOUT", 60)
)

OSCTL_MAX_TASKS = int(os.environ.get("OSCTL_MAX_TASKS", 150))

OSCTL_HEARTBEAT_PEERING_OBJECT_NAME = os.environ.get(
    "OSCTL_HEARTBEAT_PEERING_OBJECT_NAME", "rockoon.osdpl"
)

if OSCTL_HEARTBEAT_INTERVAL:

    @kopf.timer(
        "zalando.org",
        "v1",
        "kopfpeerings",
        interval=OSCTL_HEARTBEAT_INTERVAL,
        when=lambda name, **_: OSCTL_HEARTBEAT_PEERING_OBJECT_NAME == name,
    )
    def heartbeat(spec, **kwargs):
        global HEARTBEAT
        global CURRENT_NUMBER_OF_TASKS
        HEARTBEAT = time.time()
        CURRENT_NUMBER_OF_TASKS = event_queue_var.get().qsize()


# The version of curren cluster release, example 8.4.0-rc+22.1
OSCTL_CLUSTER_RELEASE = os.environ.get("OSCTL_CLUSTER_RELEASE", "")


# A dict of OpenStack node labels per role in json format.
# See the structure below
# If a node has one of specified labels with the matching value,
# an appropriate NodeWorkloadLock object will be created.
def _parse_node_roles_from_env():
    roles = json_from_env(
        "OSCTL_OPENSTACK_NODE_LABELS",
        {
            "controller": {"openstack-control-plane": "enabled"},
            "compute": {"openstack-compute-node": "enabled"},
            "gateway": {"openstack-gateway": "enabled"},
        },
    )

    parsed_roles = {const.NodeRole[k]: v for k, v in roles.items()}
    for d in parsed_roles.values():
        if len(d.keys()) != 1:
            raise kopf.PermanentError(
                "OSCTL_OPENSTACK_NODE_LABELS expect 1 label key"
            )
    return parsed_roles


try:
    OSCTL_OPENSTACK_NODE_LABELS = _parse_node_roles_from_env()
except Exception as exc:
    # may raise KeyError from Enum, AttributeError from non-dict items()
    # and whatever from json.load
    raise kopf.PermanentError(
        f"OSCTL_OPENSTACK_NODE_LABELS invalid format - {exc}"
    )


def _get_internal_labels():

    int_labels = {}
    for role_name, role_labels in OSCTL_OPENSTACK_NODE_LABELS.items():
        for label_name in role_labels.keys():
            int_labels[role_name] = {f"rockoon-{label_name}": "controller"}
    return int_labels


OSCTL_OPENSTACK_NODE_LABELS_INTERNAL = _get_internal_labels()

# The dict defining proxy settings
OSCTL_PROXY_DATA = json_from_env("OSCTL_PROXY_DATA", {"enabled": False})
OSCTL_PROXY_SECRET_NAMESPACE = os.environ.get(
    "OSCTL_PROXY_SECRET_NAMESPACE", "osh-system"
)
OSCTL_CDN_CA_BUNDLE_DATA = json_from_env("OSCTL_CDN_CA_BUNDLE_DATA", {})
OSCTL_CDN_CA_BUNDLE_SECRET_NAMESPACE = os.environ.get(
    "OSCTL_CDN_CA_BUNDLE_SECRET_NAMESPACE", "openstack"
)
OSCTL_CEPH_DEPLOYMENT_NAMESPACE = os.environ.get(
    "OSCTL_CEPH_DEPLOYMENT_NAMESPACE", "rook-ceph"
)
OSCTL_LMA_DEPLOYMENT_NAMESPACE = os.environ.get(
    "OSCTL_LMA_DEPLOYMENT_NAMESPACE", "stacklight"
)
OSCTL_TF_DEPLOYMENT_NAMESPACE = os.environ.get(
    "OSCTL_TF_DEPLOYMENT_NAMESPACE", "tf"
)


# The dict defining IAM data {"client": "os", "enabled": True, "oidcCASecret": "oidc-cert", url: "https://1.2.3.4"}
OSDPL_IAM_DATA = json_from_env("OSDPL_IAM_DATA", {"enabled": False})

# List with pod networks data [{"cidr": "1.2.3.0/24"}]
OSCTL_POD_NETWORKS_DATA = json_from_env(
    "OSCTL_POD_NETWORKS_DATA", [{"cidr": "192.168.0.0/16"}]
)

CONF = Config()


class InfiniteBackoffsWithJitter:
    def __iter__(self):
        while True:
            yield 10 + random.randint(-5, +5)


@kopf.on.startup()
def configure(settings: kopf.OperatorSettings, **_):
    settings.watching.connect_timeout = 1 * 60
    # NOTE(vsaienko): The watching.server_timeout is used to set timeoutSeconds
    # for kubernetes watching request.
    # Timeout for the list/watch call. This limits the duration of the call,
    # regardless of any activity or inactivity.
    # IMPORTANT: this timeout have to be less than aiohttp client.timeout
    settings.watching.server_timeout = os.environ.get(
        "KOPF_WATCH_STREAM_TIMEOUT", 1 * 300
    )
    # Defines total timeout for aiohttp watching session.
    settings.watching.client_timeout = 1 * 600
    # setting unique finalizer name for each controller
    settings.persistence.finalizer = (
        f"lcm.mirantis.com/{OSCTL_HEARTBEAT_PEERING_OBJECT_NAME}-finalizer"
    )
    settings.networking.error_backoffs = InfiniteBackoffsWithJitter()
    settings.execution.max_workers = CONF.getint("osctl", "max_workers")


# HELM SETTINGS
HOME = str(Path.home())
HELM_REPOSITORY_CACHE = os.environ.get(
    "HELM_REPOSITORY_CACHE", os.path.join(HOME, ".cache/helm/repository")
)
HELM_CHARTS_DIR = os.environ.get("HELM_CHARTS_DIR", "/opt/operator/charts/")
# END HELM SETTINGS
